{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')\n","\n","%cd /content/drive/MyDrive/repos/Epilepsy_Microglia\n","%pip install -q -r requirements.txt"],"metadata":{"id":"2SLBPKyxrxTS"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**3. Downstream**\n","1. scANVI rough labeling\n","2. Reference label transfer\n","\n","-----------------------------------------------\n","*Logs*\n","\n","20251102 : Run scVI on HVGs, then transfer the latent to full data"],"metadata":{"id":"saiF82oyXdeb"}},{"cell_type":"code","source":["import torch\n","\n","# Single Cell Libraries\n","import scvi\n","import scanpy as sc\n","import anndata as ad\n","\n","# Data Processing and Plotting\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","import igraph\n","import leidenalg\n","import random\n","\n","# File grab\n","import os\n","import tempfile\n","import pooch\n","import shutil, subprocess, glob\n","import gzip\n","\n","print(torch.__version__)\n","print(scvi.__version__)\n","print(torch.cuda.is_available())\n","\n","# Random key\n","SEED = 42\n","random.seed(SEED)\n","np.random.seed(SEED)\n","torch.manual_seed(SEED)\n","torch.cuda.manual_seed_all(SEED)\n","torch.backends.cudnn.deterministic = True\n","torch.backends.cudnn.benchmark = False\n","scvi.settings.seed = SEED"],"metadata":{"id":"lctTbUeseW-n"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# not for github\n","\n","# Pathway/Settings\n","SOURCEDIR = \"/content/drive/MyDrive/datas/epilepsy_microglia/processed/GSE201048/\"\n","FILE_LABEL = \"kumar_qc_20251101v3\"\n","HVG_FILE_LABEL = \"kumar_hvg_20251101v3\"\n","\n","# scVI\n","sc.set_figure_params(dpi_save=300, frameon=False)\n","sc.settings.figdir = \"/content/drive/MyDrive/repos/Epilepsy_Microglia/figures\"\n","\n","adata = sc.read_h5ad(os.path.join(SOURCEDIR,f\"{FILE_LABEL}.h5ad\"))\n","adata_hvg = sc.read_h5ad(os.path.join(SOURCEDIR,f\"{HVG_FILE_LABEL}.h5ad\"))\n","model_dir = os.path.join(SOURCEDIR,\"model_v3\")\n","model = scvi.model.SCVI.load(model_dir, adata=adata_hvg)\n","assert all(adata.obs_names == adata_hvg.obs_names)\n","\n","adata"],"metadata":{"id":"P1oxLuKOfQb8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# not for github\n","adata.write(os.path.join(SOURCEDIR, f\"{FILE_LABEL}.h5ad\"))\n","model.save(os.path.join(SOURCEDIR, \"model_v3\"), overwrite=True)"],"metadata":{"id":"ZdyDy1TVQuxQ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["os.remove(\"/content/drive/MyDrive/datas/epilepsy_microglia/processed/GSE201048/kumar_v2_20251020.h5ad\")"],"metadata":{"id":"AOt3gGxb8jpJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["SCVI_LATENT_KEY = \"X_scVI\"\n","latent = model.get_latent_representation()\n","adata.obsm[SCVI_LATENT_KEY] = latent\n","latent.shape"],"metadata":{"id":"CpAmL7OBhCOu"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["adata"],"metadata":{"id":"03aQWLXx9SFn"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["sc.pp.neighbors(adata, use_rep=\"X_scVI\")\n","sc.tl.umap(adata, random_state=SEED)\n","sc.tl.leiden(adata,\n","             resolution=0.5,\n","             flavor=\"igraph\",\n","             random_state=SEED,\n","             key_added=\"leiden_0.5_v3\")\n","sc.pl.umap(adata, color=[\"leiden_0.5_v3\"], save=\"20251101v3_scvi_leiden0.5\")\n","sc.tl.rank_genes_groups(adata, \"leiden_0.5_v3\", method=\"wilcoxon\")\n","sc.pl.rank_genes_groups(adata, n_genes=10, sharey=False,\n","                        save=\"20251101v3_scvi_leiden0.5_rank_genes_groups\")"],"metadata":{"id":"-JFTCy1T105P"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["sc.pl.umap(adata, color=[\"F13A1\"])"],"metadata":{"id":"wq443gwr-rqJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["genes_row1 = [\"SALL1\", \"HEXB\", \"CSF1R\", \"CX3CR1\"]          # microglia / homeostatic\n","genes_row2 = [\"LST1\", \"FCN1\", \"S100A8\", \"VCAN\"]            # monocyte / MDM\n","genes_row3 = [\"LYVE1\", \"CD163\", \"F13A1\"]                   # BAM / PVM\n","\n","all_genes = genes_row1 + genes_row2 + genes_row3\n","existing = [g for g in all_genes if g in adata.var_names]\n","\n","sc.pl.umap(\n","    adata,\n","    color=existing,\n","    ncols=4,\n","    frameon=False,\n","    size=15\n",")\n","plt.show()"],"metadata":{"collapsed":true,"id":"mws_tM_pBOqd"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["genes_row1 = [\"P2RY12\", \"TGFBR1\", \"TREM2\", \"TMEM119\"]\n","genes_row2 = [\"CD2\", \"CD3E\", \"NKG7\", \"GNLY\"]\n","genes_row3 = [\"PLP1\", \"MBP\", \"MOG\"]\n","\n","all_genes = genes_row1 + genes_row2 + genes_row3\n","existing = [g for g in all_genes if g in adata.var_names]\n","\n","sc.pl.umap(\n","    adata,\n","    color=existing,\n","    ncols=4,\n","    frameon=False,\n","    size=15\n",")\n","plt.show()"],"metadata":{"id":"caahra0qEwEa"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Annotating schema v0.2 (20251016)\n","\n","import scipy.sparse as sp\n","\n","def get_bool(adata, gene, thr):\n","    x = adata[:, gene].X\n","    if sp.issparse(x):\n","        mask = (x > thr)\n","        return np.ravel(mask.toarray() if mask.ndim > 1 else mask.A1)\n","    else:\n","        return np.ravel(x) > thr\n","\n","cd45_hi = get_bool(adata, 'PTPRC', 1.5)\n","cd14_hi = get_bool(adata, 'CD14', 1.0)\n","cd3_hi  = get_bool(adata, 'CD3E', 1.0)\n","cd56_hi = get_bool(adata, 'NCAM1', 1.0)\n","cd19_hi = get_bool(adata, 'CD19', 0.5)\n","cd20_hi = get_bool(adata, 'MS4A1', 0.5)\n","\n","adata.obs['manual_label_v1'] = \"Unknown\"\n","\n","# immune cell : CD45hi\n","adata.obs.loc[cd45_hi & cd14_hi, 'manual_label_v1'] = \"Mono_Mac\"\n","adata.obs.loc[cd45_hi & cd3_hi, 'manual_label_v1'] = \"T\"\n","adata.obs.loc[cd45_hi & cd56_hi & ~cd3_hi, 'manual_label_v1'] = \"NK\"\n","adata.obs.loc[cd45_hi & cd19_hi & cd20_hi, 'manual_label_v1'] = \"B\"\n","\n","def sparse_any_gt(adata, genes, thr):\n","    X = adata[:, genes].X\n","    if not sp.issparse(X):\n","        return (X > thr).any(axis=1)\n","    ## keep as sparse boolean matrix\n","    return np.array((X > thr).sum(axis=1)).ravel() > 0\n","\n","# microglia: CD45lo but microglia RNA genes\n","mg = (~cd45_hi) & sparse_any_gt(adata, ['P2RY12', 'TMEM119'], 0.5)\n","adata.obs.loc[mg, 'manual_label_v1'] = 'Microglia'\n","\n","# oligo (RNA)\n","oligo = (~cd45_hi) & sparse_any_gt(adata, ['PLP1','MBP','MOG'], 0.5)\n","adata.obs.loc[oligo, 'manual_label_v1'] = 'Oligo'"],"metadata":{"id":"h-QXvwG_w7KY"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(pd.value_counts(adata.obs[\"manual_label_v1\"]))"],"metadata":{"id":"dFD_oUbG9djN"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# 2. scANVI label stabilization\n","\n","scvi.model.SCANVI.setup_anndata(\n","    adata,\n","    labels_key=\"manual_label_v1\",\n","    unlabeled_category=\"Unknown\",\n","    batch_key=\"sample_id\"\n",")\n","\n","sanvi = scvi.model.SCANVI.from_scvi_model(\n","    model,\n","    adata=adata,\n","    unlabeled_category=\"Unknown\",\n","    labels_key=\"manual_label_v1\"\n",")\n","\n","sanvi.train(\n","    max_epochs=200,\n","    check_val_every_n_epoch=10,\n","    n_samples_per_label=100\n",")\n","adata.obs['label_scanvi'] = sanvi.predict()"],"metadata":{"id":"m7tTZ8BjaSmV","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Check results\n","\n","sc.pp.neighbors(adata, use_rep=\"X_scVI\")\n","sc.tl.umap(adata, random_state=SEED)"],"metadata":{"id":"zGAPuSILaDQG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["sc.pl.umap(adata, color=[\"label_scanvi\"])"],"metadata":{"id":"IFJHwAzg0_Nw"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["sc.pl.umap(adata, color=[\"PTPRC\"])"],"metadata":{"id":"gWqYi-hTjFef"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["\"PTPRC\" in adata.var_names"],"metadata":{"id":"k3ii_ofVbaSr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["sc.pp.neighbors(adata, use_rep=\"X_pca\", random_state=SEED)\n","sc.tl.umap(adata, random_state=SEED)\n","sc.pl.umap(adata, color = [\"label_scanvi\"])"],"metadata":{"id":"hbeiokHLH6VB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# @title\n","# schema v0.1 (20251010) Dumped 20251012\n","\n","annotation_schema = {\n","    \"version\": \"0.1\",\n","    \"levels\": {\n","        \"L0_major\": {\n","            \"Microglia\": {\n","                \"positive\": [\"P2RY12\", \"TMEM119\", \"CX3CR1\", \"TREM2\", \"CSF1R\"],\n","                #\"negative\": [\"PTPRC\", \"MBP\", \"AQP4\", \"CLDN5\", \"VWF\", \"SLC17A7\", \"GAD1\", \"MOG\"],\n","                \"scores\": [\n","                    {\n","                        \"module\": \"PIC_core\",  # e.g., [AIF1, TYROBP, CYBB, NFKBIA, IL1B, CCL3...]\n","                        \"use_for_subtype\": True,\n","                    }\n","                ],\n","            },\n","            \"Astro\": {\n","                \"positive\": [\"GFAP\", \"AQP4\", \"ALDH1L1\", \"S100B\", \"SOX9\"],\n","                #\"negative\": [\"P2RY12\", \"MBP\", \"MOG\", \"CLDN5\", \"KDR\"],\n","            },\n","            \"Oligo\": {\n","                \"positive\": [\"PLP1\", \"MBP\", \"MOG\"],\n","                #\"negative\": [\"AQP4\", \"P2RY12\"],\n","            },\n","            \"Endo\": {\n","                \"positive\": [\"CLDN5\", \"KDR\", \"FOXF2\"],\n","                #\"negative\": [\"P2RY12\", \"MBP\"],\n","            },\n","            \"Pericyte\": {\n","                \"positive\": [\"CLDN5\", \"KDR\", \"FOXF2\"],\n","                #\"negative\": [\"P2RY12\", \"MBP\"],\n","            },\n","            \"Mono/Macro\": {\n","                \"positive\": [\"LST1\", \"PTPRC\", ],\n","                \"negative\": []\n","            },\n","            \"Immune_T_NK\": {\n","                \"positive\": [\"CD2\", \"GNLY\", \"NKG7\"],\n","                \"negative\": [\"P2RY12\"],\n","            },\n","            \"B/Plasma\": {\n","\n","            },\n","    },\n","    \"rules\": {\n","        \"scoring\": {\n","            \"method\": \"rank_pct\",  # percentile ranks, not raw cutoffs\n","            \"min_pos_pct\": 0.90,\n","            \"max_neg_pct\": 0.50,\n","        },\n","        \"doublet_flags\": [\n","            {\"and\": [\"P2RY12\", \"GNLY\"]},  # Microglia × T/NK co-expression\n","            {\"and\": [\"P2RY12\", \"MBP\"]},   # Microglia × Oligo\n","          ],\n","        },\n","    }\n","}\n","\n","# check if anything out of adata.var_names\n","L0_major = annotation_schema[\"levels\"][\"L0_major\"]\n","\n","for celltype, panel in L0_major.items():\n","    pos = panel.get(\"positive\", [])\n","    neg = panel.get(\"negative\", [])\n","    all_genes = pos + neg\n","    missing = [g for g in all_genes if g not in adata.var_names]\n","    if missing :\n","        print(f\"{celltype}: missing {len(missing)} -> {missing}\")"],"metadata":{"id":"eQwGpPr28pFM","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# @title\n","# Annotation 20251010, dumped 20251012\n","\n","# 1. Annotate the cells based on the labeling panel\n","\n","# anndata is normalized and log-transformed\n","rough_labels = np.full(adata.n_obs, \"Unknown\", dtype=object)\n","\n","L0_major = annotation_schema[\"levels\"][\"L0_major\"]\n","\n","# (1) Build a center assignment based on ( >= 2 genes, >= log1p expr 2 )\n","\n","for celltype, panel in L0_major.items():\n","    pos_genes = [g for g in panel.get(\"positive\", []) if g in adata.var_names]\n","    if len(pos_genes) == 0:\n","        continue\n","\n","    # boolean mark for each cell x each gene (pos)\n","    expr = adata[:, pos_genes].X\n","    if not isinstance(expr, np.ndarray):\n","      expr = expr.toarray()\n","\n","    # set expression threshold\n","    threshold = 2.0\n","\n","    # count how many genes exceed threshold per cell\n","    mark = (expr >= threshold).sum(axis=1)\n","    mask = mark >= 2\n","    rough_labels[mask] = f\"{celltype}\"\n","\n","adata.obs[\"rough_annot_center\"] = rough_labels\n","print(pd.value_counts(adata.obs[\"rough_annot_center\"]))\n","\n","# (2) Build PCA neighbors\n","sc.pp.pca(adata, n_comps=30)\n","sc.pp.neighbors(adata, use_rep=\"X_pca\", n_neighbors=15)\n","\n","# (3) Apply neighborhood purity filter : rule out > 50% other labels.\n","n_neighbors = adata.uns[\"neighbors\"][\"params\"][\"n_neighbors\"]\n","connectivities = adata.obsp[\"connectivities\"]\n","\n","# for each cell, compute fraction of neighbors with same label\n","labels = adata.obs[\"rough_annot_center\"].values\n","pure = np.full(len(labels), True)\n","\n","for i in range(len(labels)):\n","    if pd.isna(labels[i]):\n","        continue\n","    # neighbor indices and weights\n","    neigh_idx = connectivities[i].indices\n","    neigh_labels = labels[neigh_idx]\n","    same = np.sum(neigh_labels == labels[i])\n","    valid = np.sum(~pd.isna(neigh_labels))\n","    if valid > 0 and same / valid < 0.5:\n","        pure[i] = False\n","\n","adata.obs.loc[~pure, \"rough_annot_center\"] = \"Unknown\"\n","\n","# summary\n","print(pd.value_counts(adata.obs[\"rough_annot_center\"]))"],"metadata":{"id":"DRPD7lOrPgOe","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# @title\n","# not for github\n","# check double-positive cell proportion\n","\n","import numpy as np\n","import scipy.sparse as sp\n","\n","x = adata[:, \"MS4A1\"].X\n","y = adata[:, \"CD19\"].X\n","\n","if sp.issparse(x): x = x.toarray()\n","if sp.issparse(y): y = y.toarray()\n","\n","x = np.ravel(x)\n","y = np.ravel(y)\n","\n","# Correct boolean combination\n","bcell_pos = (x > 0.5) & (y > 0.5)\n","\n","n_pos = int(bcell_pos.sum())\n","print(f\"B-cell double positive: {n_pos} cells\")\n"],"metadata":{"cellView":"form","id":"nBkG0kQR55X6"},"execution_count":null,"outputs":[]}]}